### ISSUE TITLE
unable to drain k8s node running Istio components with PDB

### ISSUE THREAD

**Issue by jansmets** [istio/istio#12602]
Deployed istio 1.1.0 RC with helm chart (not template)

I would like to drain out a specific node because it needs a reboot. 
The PODs
```
Annotations:        scheduler.alpha.kubernetes.io/critical-pod:
```
prevents that.

```
error when evicting pod "istio-policy-d48655744-s8ggt" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

(also note that annotation scheduler.alpha.kubernetes.io/critical-pod is deprecated and will be removed in k8s 1.14!)

**Comment by dwradcliffe** [comment]
I don't think this is related to the critical-pod annotation, but instead the PodDisruptionBudget.

You can view with `kubectl describe pdb istio-policy -n istio-system`.
I'm guessing you only have 1 pilot pod running, and the PDB is preventing the eviction because that would violate the budget.

Try running 2 replicas of pilot.

**Comment by linsun** [comment]
agreed, try scale out istio-policy first before drain the node as PDB requires 1 replica to be alive.

**Comment by linsun** [comment]
@jansmets is this still an issue?

**Comment by Lysholm** [comment]
The default helm installs are set up with 1 replica and a PDB for several deployments / HPAs. This seems like a case of a bad default config. PDBs are a feature for HA deployments, and HA deployments imply 2+ replicas.

The reason this is bad is that it breaks `kubectl drain`. In itself not good but automatic node-pool upgrades may also fail or take unacceptably long to apply. Node-pools will automatically upscale but then may fail to downscale as well.

Don't know if this is a feature or a bug but it's not a great default config for istio installs. I would either remove the PDBs or increase default replicas to 2.

Affected deployments (spec.replicas = 1): certmanager, istio-galley
Affected HPA (spec.minReplicas = 1): istio-egressgateway, istio-ingressgateway, istio-policy, istio-telemetry, istio-pilot

**Comment by kinihun** [comment]
Is there any plan to resolve this issue. It's still a blocker when using such a tool like kured. It results in a node being effectively locked out of action almost indefinitely.

**Comment by johscheuer** [comment]
We also ran into this issue and worked around it by setting the `replicaCount` for all components greater than 1. 

Actually resolving this issue has some challenges from my which i'm not sure how to solve best:

We want to create a `PodDisruptionBudget` only in the following cases:

- `replicaCount` is set to 1 or greater
- When `PodDisruptionBudget` is activated obviously 

but what if you don't set the `replicaCount`  because you use the autoscaling feature with `minReplicas: 1`? Should the `pdb` created since there can be cases where more than 1 replica is available but if there is no load on your cluster you will have the same challenges as before.

Probably a quick win would be to allow setting the `pdb` behaviour on components e.g. you can globally activate `pdb` but deactivate if for specific components.

**Comment by linsun** [comment]
Think this is still a valid issue for the new installer/operator.

The resulting is we produce a production high availability profile that have multiple replicas and specifically targeted to resolve upgrade without down time prob.   We don't want this to be a solution just for istio-policy as istio-policy is going away soon.

How does that sound @johscheuer @jansmets @dwradcliffe?

cc @howardjohn @sdake @ostromart

**Comment by Arnold1** [comment]
@dwradcliffe  @linsun running in the same issue:

```
$ kops rolling-update cluster --yes
...
error when evicting pod "istio-policy-578bcb878f-2dmlw" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pod "istio-galley-84749d54b7-24qrp" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pod "istio-policy-578bcb878f-2dmlw" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pod "istio-galley-84749d54b7-24qrp" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

```$ kubectl describe pdb istio-policy -n istio-system
Name:           istio-policy
Namespace:      istio-system
Min available:  1
Selector:       app=policy,istio=mixer,istio-mixer-type=policy,release=istio
Status:
    Allowed disruptions:  0
    Current:              1
    Desired:              1
    Total:                1
Events:                   <none>
```

what to do to fix it?

**Comment by vigohe** [comment]
@Arnold1 you need to add one more replica for istio-policy & istio-galley.

If you did install istio with helm you need to add the following lines to values.yaml

```
galley:
  enabled: true
  replicaCount: 2

security:
  enabled: true
  replicaCount: 2
```

Then apply it:

```
helm upgrade --install istio istio.io/istio --namespace istio-system -f values.yaml
```

**Comment by colincoghill** [comment]
citadel has the same problem in istio 1.4, except you don't appear to be able to specify its replicaCount, that I can see.

**Comment by vigohe** [comment]
@colincoghill for citadel it's the security section.

values.yaml
```
security:
  enabled: true
  replicaCount: 2
```

**Comment by colincoghill** [comment]
aha, brilliant, that worked for me, thank you @vigohe

**Comment by Arnold1** [comment]
@vigohe I get this error:
```$ helm upgrade --install istio istio.io/istio --namespace istio-system -f values.yaml
Error: failed to download "istio.io/istio" (hint: running `helm repo update` may help)```

**Comment by Arnold1** [comment]
what worked is:
cd istio-1.4.0
and than edit install/kubernetes/helm/istio/values.yaml:
```
  defaultPodDisruptionBudget:
    enabled: false
```

and than:
```
helm upgrade --install istio install/kubernetes/helm/istio --namespace istio-system -f install/kubernetes/helm/istio/values.yaml
```

is that the recomended way for istio 1.4?

**Comment by kovalyukm** [comment]
So, what is Istio recommended solution to correctly drain node with any istio component?

Istio 1.4.2 installed with istioctl has these values for prod (default) profile:

k get pdb -n istio-system 
NAME                     MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
certmanager              1               N/A               0                     12d
ingressgateway           1               N/A               0                     12d
istio-citadel            1               N/A               0                     12d
istio-galley             1               N/A               0                     12d
istio-pilot              1               N/A               0                     12d
istio-policy             1               N/A               0                     12d
istio-sidecar-injector   1               N/A               0                     12d
istio-telemetry          1               N/A               0                     12d

ALLOWED DISRUPTIONS = 0 for all components.

Thanks!

**Comment by tariq1890** [comment]
Re-opening since this was auto-closed and we have yet to receive a working solution to this.

I think we should have a good set of defaults here. As per the manifests, the `replicaCount` is set to 1 (the default value) if `autoscaling` is enabled. If a replicaCount of 1 leads to a problematic PDB, then this default value should certainly be reconsidered

**Comment by tariq1890** [comment]
@linsun @howardjohn Requesting your attention on this. I think we need better defaults here. Since we are in `istiod`, I'd just suggest defaulting the replicaCount of  istiod and the gateways to 2. This seems to work rather well especially with PDB's being enabled by default

**Comment by lucbas** [comment]
Is also a pain for us and offer my help wherever required.

**Comment by mandarjog** [comment]
@ostromart you changed the defaults recently to have replica counts =2, right? 
is this resolved by that fix?

**Comment by ostromart** [comment]
Only internally, we've discussed doing it in OSS but there are some challenges there. Specifically, we should only set 2 if high availability is selected (and PDBs installed) and this should be dynamically checked. Not hard, just needs some cycles and it's not been a high priority. If you're interested in taking a look @lucbas i can offer guidance.

**Comment by kramvan1** [comment]
For Managed Istio, replicas is set to 2 and PDB is 1, so node maintenance should be allowed as normal.
```
k get pdb -A
NAMESPACE      NAME                    MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
istio-system   cluster-local-gateway   1               N/A               1                     71m
istio-system   istio-egressgateway     1               N/A               1                     44h
istio-system   istio-ingressgateway    1               N/A               1                     44h
istio-system   istiod                  1               N/A               1                     44h
```

**Comment by kriston13** [comment]
Hi,
We're running into this as well. We're using GKE's preemptible node pools, so we need to safely drain an "old" node in order to prepare it to be pre-empted. But... The istio items are causing us issues. Is there any guidance on how to do this with the istio operator? Thanks!

**Comment by soakr** [comment]
I faced same issue around istio-ingressgateway.
After several try, maybe I found some  solution for this issue even if it's not smart.
In my case, the most critical issue was lack of machine resources.  There were no nodes to deploy new pod to fill demanded number of pods.

Since lack of machine resource, I'm not able to check whether it works for now but I'm going to introduce it.
We can get actual setting for each resource by using like following commands.

**deployment / istio-ingressgateway**
```shell
kubectl get deployment istio-ingressgateway -n istio-system -o yaml > dep.yaml
```

**pdb / ingressgateway**
```shell
kubectl get pdb ingressgateway -n istio-system -o yaml > pdb.yaml
```

then we can increase count of replicas or something like that, and deploy k8s to use `apply -f` command.

Hope this can help you.

**Comment by tvkit** [comment]
In my case, manually zeroing the following replicas permitted `kops` to complete its ig revisions

```
k scale deploy istiod --replicas=0
k scale deploy istio-ingressgateway --replicas=0
k scale deploy istio-egressgateway --replicas=0

```

**Comment by sdake** [comment]
Changing this to P0. @linsun if you don't have time to work on this prior to 1.7, would you mind unassigning yourself?

I am unclear what the actual problem is since this issue occurred in Istio 1.3. It sounds like there are two problems here:

- If not enough nodes are available to permit an eviction, an eviction doesn't occur (this seems normal and expected behavior).
- There is no way to drain an active node.

@tariq,

What is the desired behavior here?

Having a HA mode in Istio which does the following:
1. installs PDBs
1. increases replicas to 2

If that is the target implementation, should be straightforward.

@ostromart please let me know if this matches what you have done elsewhere, and I'll address the work with a PR.

Cheers,
-steve

**Comment by sdake** [comment]
@tariq1890 ^

**Comment by tariq1890** [comment]
Sorry for the late response. I think a default of having 2 replicas with PDB installed is good. That's how we are running istio in our kubernetes clusters.

**Comment by JimmyCYJ** [comment]
Confirmed with @linsun this could be closed.

**Comment by tariq1890** [comment]
Why close this? I don't see any concrete resolution provided for this

**Comment by linsun** [comment]
Sorry I was pinged about this issue while on mobile phone... because istio-policy pod is getting deprecated in 1.7, my thinking is that we can close this out for 1.7.  Think it is also non default for 1.6 either.

**Comment by sdake** [comment]
Hi gang. This is a duplicate of actions taking place at https://github.com/istio/istio/issues/24000. Unfortunately, much of the context is here, however, I'd encourage the PR work taking place in the above issue to account for the community's responses here.

Cheers,
-steve

**Comment by sazzle2611** [comment]
Hi all, I have this same issue only it is on a kubernetes cluster (v 17.2) with only 2 nodes, one master and one worker, meaning as pods are only scheduled on the worker then we can't get around this problem by autoscaling.

He don't need HA but it would be nice to be able to drain the node properly, is there any way to disable the pdb (and hpa) for the ingress gateways and istiod?

We are using istio 1.6 installed with istioctl and this is the current ingress gateway config but it still creates pdb

  components:
    ingressGateways:
    - name: istio-ingressgateway
      enabled: true
      k8s:
        replicaCount: 1
        hpaSpec:
          minReplicas: 1
          maxReplicas: 1

If I remove the hpaSec then it sets a hpa with max replicas 5 which is definitely not what we want.

Thanks,
Sarah

**Comment by davidkarlsen** [comment]
This can be avoided with the following spec:
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio-controlplane
  namespace: istio-system
spec:
  addonComponents:
    kiali:
      enabled: true
    prometheus:
      enabled: false
    tracing:
      enabled: true
  components:
    cni:
      enabled: true
    ingressGateways:
    - enabled: true
      k8s:
        hpaSpec:
          minReplicas: 2
      name: istio-ingressgateway
    pilot:
      enabled: true
      k8s:
        hpaSpec:
          minReplicas: 2
  meshConfig:
    enablePrometheusMerge: true
  profile: default
  values:
    cni:
      excludeNamespaces:
      - istio-system
      - kube-system
      logLevel: info
    kiali:
      dashboard:
        grafanaURL: http://monitoring-grafana.monitoring.svc
      prometheusAddr: http://monitoring-prometheus-server.monitoring.svc
```
which should probablt be default (the hpa values)

**Comment by bianpengyuan** [comment]
Related: https://github.com/istio/istio/issues/24000

**Comment by Vaccano** [comment]
> This can be avoided with the following spec:
> 
> ```
> apiVersion: install.istio.io/v1alpha1
> kind: IstioOperator
> metadata:
>   name: istio-controlplane
>   namespace: istio-system
> spec:
>   addonComponents:
>     kiali:
>       enabled: true
>     prometheus:
>       enabled: false
>     tracing:
>       enabled: true
>   components:
>     cni:
>       enabled: true
>     ingressGateways:
>     - enabled: true
>       k8s:
>         hpaSpec:
>           minReplicas: 2
>       name: istio-ingressgateway
>     pilot:
>       enabled: true
>       k8s:
>         hpaSpec:
>           minReplicas: 2
>   meshConfig:
>     enablePrometheusMerge: true
>   profile: default
>   values:
>     cni:
>       excludeNamespaces:
>       - istio-system
>       - kube-system
>       logLevel: info
>     kiali:
>       dashboard:
>         grafanaURL: http://monitoring-grafana.monitoring.svc
>       prometheusAddr: http://monitoring-prometheus-server.monitoring.svc
> ```
> 
> which should probablt be default (the hpa values)

It is important to note that this solution only works if you installed Istio using the Istio Operator. (At least that was my experience.)

**Comment by Kostanos** [comment]
Just scaling **istiod** deployment to 2 and after scaling it back to 1 worked for me.
Meanwhile the draining node is in cordoned state.
I'm using lens for it, but here is a kubectl:
```bash
kubectl cordon <your node to drain>
kubectl scale --replicas=2 deployemnt/istiod
# Make sure it was scaled, when you scale it back, it will remove pod from cordoned node
kubectl scale --replicas=1 deployment/istiod
kubectl drain <your node to drain>
```

PS. in my case, istiod was in separate node from other istio related services, and it worked.

**Comment by dozer75** [comment]
@Kostanos suggestion helped me during the AKS outage yesterday due to the Ubuntu 18.04 issues to heal the cluster. In addition I had to run the same command on `istio-ingressgateway` as well.

**Comment by PrincePranav235** [comment]
Another solution also will be to disable the pdb or make min available to 0.

**Comment by dgn** [comment]
this seems like a fairly big problem with our default settings. if the minAvailable in the PDB is hard-coded, we'll need to stop creating it for deployments where our replicaCount < 2 IMO

**Comment by Stevenjin8** [comment]
resolved by #56630

**Comment by FilipB** [comment]
@Stevenjin8 This issue is not resolved by https://github.com/istio/istio/pull/56630, could you please re-open?
https://github.com/istio/istio/pull/56630 only allows to configure `minAvailable` in PDB but it does not change the default value which is still `1`.  With `minAvailable = 1` (default value) in PDB and `minReplicas = 1` (default value) in HPA, the issue is still visible.

**Comment by Stevenjin8** [comment]
I'll reopen and try to summarize the issue. With #56630, users can now configure their PDB and HPA so that nodes running istio pods (istiod) can be drained. The problem is the default installation and the question is: should the default install of istiod configure istiod to be HA? If so we should set the PDB `minAvailable = 1`, but by default deploy two istiods. If we don't want HA, then we can keep istiod as is, but should change PDB `minAvailable` to 0. @FilipB is that an accurate summary?

I am hesitant to change things just because we don't know if any users, for whatever reason, rely on the default behavior, and for impacted users, they have a reasonable workaround.

**Comment by FilipB** [comment]
I went through all earlier attempts to fix this issue and here is what I found:

- the cleanest option is to change current k8s behavior as described in https://github.com/kubernetes/kubernetes/issues/93476. With that fix, there is no need to change anything on Istio side.
- without the k8s behavior change above,  there are following options:

  1.  do not create PDB (or set `minAvailable = 0`) if `minReplicas < 2` in HPA - this seems to be the best option as the original reason to introduce PDB was to fix https://github.com/istio/istio/issues/9951 and the behavior during chart upgrade might be different now showing no downtime even without PBD (~to be confirmed~), UPDATE: Confirming that new replica is started before old replica is terminated even without PDB during helm upgrade of istiod.
  1.  set `minReplicas = 2` in HPA if PDB exists - default profile always creates PDB so there would be 2 replicas by default. Arguments against 2 replicas by default were: resource consumption, more complicated debugging
  1. create new HA profile - arguments against were: It does not make sense to create another profile to change just a few default values

The best option for now seems to be option `i.` which should be reverted if https://github.com/kubernetes/kubernetes/issues/93476 ever gets resolved.
